### DISCUSSION TITLE
Pyinstaller cannot solve the environments problem of the inner program in the code file

### DISCUSSION THREAD

**Discussion by haiw2sm** [pyinstaller/pyinstaller#8366] [Help]
I writing a project about gui with pyqt, socket(there are two independent program called client and server in a project，which been created by the virtualenv, to realize the local communication). I mostly used the following spec file code to pack the project: 
```
# -*- mode: python ; coding: utf-8 -*-
import os
from pathlib import Path
from PyInstaller.archive.pyz_crypto import PyiBlockCipher

def find_py_files(directory):
    """
    Traverse the project directory，find out all the py file
    """
    pys = [] # store the py file found
    for root, dirs, files in os.walk(directory):
        for file in files:
            path = os.path.join(root, file)
            if not file.endswith('.py'):
                continue
            pys.append(path)

    return pys


py_files = find_py_files('{the project absolute path}')

# avoid being decompiled
block_cipher = PyiBlockCipher(key='12345')

a = Analysis(
    py_files,
    pathex=[],
    binaries=[],
    datas=[],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    cipher=block_cipher,
    noarchive=False,
)
```
There is no problem with the execution of the packaged entry program（also is the server socket program）,But the problem I'm having is the client program, which I started with QProcess in pyqt， the core partial code is the following:
```python
p = QProcess()
p.start('start.cmd', ['param1', 'param2'])
```
the outer command file is:
```cmd
@echo off
chcp 65001
set PYTHONPATH={Project Module Root}/venv/Scripts
python -m {Project Module Root}.client -u %1 -m %2
chcp 936
```

- [ x ] The logs show that the modules written in the project are missing：

```text
2024-03-20 21:32:26  "core\controllers\qProcessManager.py:39" -<lambda>-[print]-  \.pyenv\pyenv-win\versions\3.11.0\python.exe: Error while finding module specification for '{project module root}.{subModule}client' (ModuleNotFoundError: No module named 'subModule')
```

❓ : Do I need to set the PYTHON environment variable to the python interpreter in the dist directory in the QProcess or cmd file?

**Comment by rokm** [comment]
> ❓ : Do I need to set the PYTHON environment variable to the python interpreter in the dist directory in the QProcess or cmd file?

If you set it in the QProcess, then your cmd file would override it, wouldn't it?

**Reply by haiw2sm** [reply]
Yeah, i know. but now i dont know how to set it in QProcess, After all, pyinstaller is packaged in the form of Python's dynamic link library，I can't figure out how to set up the environment and python interpreter that the socket client depends on

**Reply by haiw2sm** [reply]
The socket client also depends on the modules in the project, but the modules in the project have been packaged into the exe by pyinstaller, and now I want to determine if I can call the python dynamic link library in the release version packaged by pyinstaller, because I think there will be an environment in this link library, if you use it, it is possible by default, But I don't know how to change the python executable in cmd or qprocess to python dynamic link library

**Reply by rokm** [reply]
PyInstaller uses embedded python interpreter, so you cannot use modules packaged in PyInstaller-frozen application with *external* python interpreter (`python.exe`). 

Plus, if you need external python interpreter, there is not much point in using PyInstaller in the first place. Because that external interpreter's environment will need to provide dependencies for your client script (e.g., pyqt). I see you are trying to work around this by collecting .py modules from your project, but then you would also need to do the same for all dependencies.

**Reply by haiw2sm** [reply]
OK fine，I'm going to try to package the client and the server-side program separately, and then call the executable to start

**Reply by haiw2sm** [reply]
A log snippet packaged by pyinstaller after an executable program fails to start
```
[9100] LOADER: OK.
[9100] LOADER: Manually flushing stdout and stderr
[9100] LOADER: Cleaning up Python interpreter.
```
But I always feel that there is a way there, which is to call the python dynamic link library in the version packaged and released by pyinstaller, because the principle of this link library should be similar to the virtual environment, you can find all the dependencies in the current project that are packaged, and I will read the source code when I have time to see if the dynamic link library is called to execute the external py program, whether it carries the dependency environment

**Reply by rokm** [reply]
> But I always feel that there is a way there, which is to call the python dynamic link library in the version packaged and released by pyinstaller, because the principle of this link library should be similar to the virtual environment, you can find all the dependencies in the current project that are packaged, and I will read the source code when I have time to see if the dynamic link library is called to execute the external py program, whether it carries the dependency environment

Well, good luck with that...
